pkg/executor/adapter.go

```diff
@@ -504,21 +505,25 @@ func (a *ExecStmt) Exec(ctx context.Context) (_ sqlexec.RecordSet, err error) {
 		sctx.GetSessionVars().MemTracker.SetBytesLimit(sctx.GetSessionVars().StmtCtx.MemQuotaQuery)
 	}
 
-	e, err := a.buildExecutor()
-	if err != nil {
-		return nil, err
-	}
-	// ExecuteExec will rewrite `a.Plan`, so set plan label should be executed after `a.buildExecutor`.
-	ctx = a.observeStmtBeginForTopSQL(ctx)
+	// must set plan according to the `Execute` plan before getting planDigest
+	a.inheritContextFromExecuteStmt()
 	if variable.EnableResourceControl.Load() && domain.GetDomain(sctx).RunawayManager() != nil {
 		stmtCtx := sctx.GetSessionVars().StmtCtx
 		_, planDigest := GetPlanDigest(stmtCtx)
-		_, digest := stmtCtx.SQLDigest()
-		stmtCtx.RunawayChecker = domain.GetDomain(sctx).RunawayManager().DeriveChecker(sctx.GetSessionVars().StmtCtx.ResourceGroupName, stmtCtx.OriginalSQL, digest.String(), planDigest.String())
+		_, sqlDigest := stmtCtx.SQLDigest()
+		stmtCtx.RunawayChecker = domain.GetDomain(sctx).RunawayManager().DeriveChecker(sctx.GetSessionVars().StmtCtx.ResourceGroupName, stmtCtx.OriginalSQL, sqlDigest.String(), planDigest.String())
 		if err := stmtCtx.RunawayChecker.BeforeExecutor(); err != nil {
 			return nil, err
 		}
 	}
+	a.initKvExecCounter()
```

Comment:
The most risky part of this PR. I adjusted the order of initialization of multiple fields.
1. The `RunawayChecker`
2. The `KvExecCounter`
3. The plan from `Execute` plan

The `plan` should be initialized before `KvExecCounter` and `RunawayChecker`, to use the correct digest. However, previously the `plan` for `Execute` statements are extracted inside `buildExecutor` (from `ExecuteExec`). I modifed to build the `plan` from `*plannercore.Execute`, and initialize them before `buildExecutor`.

I hope everything needed by executors can be initialized before `buildExecutor`. However, it's not true now, a lot of fields are initialized after `buildExecutor` (by setting fields of `StmtCtx`), and the dependency between them are not clear.